core/src/main/java/nl/nn/adapterframework/pipes/WsdlXmlValidator.java

```diff
@@ -197,8 +197,44 @@ public void configure() throws ConfigurationException {
 		super.configure();
 	}
 
+	@Override
+	protected PipeForward validate(Message messageToValidate, PipeLineSession session, boolean responseMode, String messageRoot) throws XmlValidatorException, PipeRunException, ConfigurationException {
+		String soapAction = session.get(SoapBindingConstants.SOAP_ACTION, "");
+		if(StringUtils.isNotEmpty(soapAction)) {
+			String soapBodyFromSoapAction = getSoapBodyFromSoapAction(soapAction, responseMode);
+			if(StringUtils.compare(messageRoot, soapBodyFromSoapAction) != 0) {
+				log.debug("messageRoot ["+soapBodyFromSoapAction+"] is determined from soapAction ["+soapAction+"]");
+				messageRoot = soapBodyFromSoapAction;
+			}
+		}
+		return super.validate(messageToValidate, session, responseMode, messageRoot);
+	}
+
+	private String getSoapBodyFromSoapAction(String soapAction, boolean responseMode) throws PipeRunException {
+		Map<QName, Binding> bindings = definition.getBindings();
+		for (Entry<QName, Binding> binding : bindings.entrySet()) {
+			List<BindingOperation> operations = binding.getValue().getBindingOperations();
+			for (BindingOperation bindingOperation : operations) {
+				List<ExtensibilityElement> elements = bindingOperation.getExtensibilityElements();
+				if(elements != null && !elements.isEmpty()) {
+					for (ExtensibilityElement element : elements) {
+						if(element instanceof SOAPOperation) {
+							String soapActionFromDefinition = ((SOAPOperation) element).getSoapActionURI();
+							if(soapActionFromDefinition.equals(soapAction)) {
+								javax.wsdl.Message message = responseMode ? bindingOperation.getOperation().getOutput().getMessage() : bindingOperation.getOperation().getInput().getMessage();
+								Map<String, Part> parts = message.getParts();
+								return parts.values().stream().map(Part::getElementName).map(QName::getLocalPart).collect(Collectors.joining(","));
+							}
+						}
+					}
+				}
+			}
+		}
+		return getSoapBody();
```

Comment:
in responseMode, this should be outputSoapBody.
However, I think it would be better to return null if no element is found, and then to stick to the original messageRoot argument of validate()